lib/dry/schema/path.rb

```diff
@@ -60,84 +60,33 @@ def initialize(keys)
 
       # @api private
       def to_h(value = EMPTY_ARRAY.dup)
-        curr_idx = 0
-        last_idx = keys.size - 1
-        hash = EMPTY_HASH.dup
-        node = hash
-
-        while curr_idx <= last_idx
-          node =
-            node[keys[curr_idx]] =
-              if curr_idx == last_idx
-                value.is_a?(Array) ? value : [value]
-              else
-                EMPTY_HASH.dup
-              end
-
-          curr_idx += 1
-        end
+        value = [value] unless value.is_a?(Array)
 
-        hash
+        keys.reverse_each.reduce(value) { |result, key| {key => result} }
```

Comment:
Any chance you could compare this wonderful one-liner with the horrible `while` loop? I'm asking because there are places like this where these procedural loops are used purely due to performance reasons. I don't remember exactly but I'm quite certain that in this case the difference was huge. This hash is produced when generating messages and it's literally the biggest performance bottleneck in dry-schema (thus in dry-validation too).

Reply:
```ruby
require "benchmark/ips"
require 'dry-schema'


path = Dry::Schema::Path.new(Array.new(1000) { ['a'..'z'].sample })

Benchmark.ips do |x|
  x.report('old impl') { path.old_to_h }
  x.report('new impl') { path.to_h }

  x.compare!
end
```

```
Warming up --------------------------------------
            old impl    70.000  i/100ms
            new impl    85.000  i/100ms
Calculating -------------------------------------
            old impl    764.275  (± 2.1%) i/s -      3.850k in   5.039660s
            new impl    875.548  (± 2.3%) i/s -      4.420k in   5.050804s

Comparison:
            new impl:      875.5 i/s
            old impl:      764.3 i/s - 1.15x  (± 0.00) slower
```
/shrug (3.0.1 & 2.7.2 looks similar)

I tried to find out where exactly `Path` operations is a hot path during actual schema validation or error/to_h generation, but failed to do so, thus I doubt that rather minor performance changes are meaningful in the real life anyway.

Reply:
In regard to message generation:
```ruby
require 'benchmark/ips'
require 'dry-schema'

PersonSchema = Dry::Schema.Params do
  required(:foo).hash do
    required(:bar).hash do
      required(:name).value(:string)
      required(:age).value(:integer, gteq?: 18)
      required(:email).value(:string)
    end
  end
end

INVALID_INPUT = {
  foo: {
    bar: {
      name: :John,
      age: "17",
      email: nil
    }.freeze
  }.freeze
}.freeze

1_000_000.times { PersonSchema.(INVALID_INPUT).errors.to_h }
```
Says something like this to me
![flamegraph](https://gist.githubusercontent.com/ojab/d6fc29ea1d8c64ecf7be9897c194b31b/raw/3da621afe08a25975bd2094348dd02d87fef799e/rbspy-2021-06-09-p0qtC3WAse.flamegraph.svg)

Reply:
When checking `result.error?(foo: {bar: :name})` in the loop:
![flamegraph](https://gist.githubusercontent.com/ojab/a86f12eef5767de1d7f84d8b28a4df4d/raw/8e855b774c94dc4b85be1fe7360d1ac56d27b8d8/rbspy-2021-06-09-gXKcXcrnYy.flamegraph.svg)